### PR TITLE
Code changes for bc policy initialisation

### PR DESCRIPTION
**Pull request by indira-17**

### CODE REVIEW

File: config/default/reppo_maniskill.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 defaults:
   - algorithm: reppo
   - env: maniskill
-  - logging: wandb_offline
+  - logging: wandb_online
```
Comment (cvoelcker) [inline config/default/reppo_maniskill.yaml:4]: try only changing the default configs if necessary. While this is potentially a harmless change, it is good practice to only touch things that really need to be changed.

File: config/logging/wandb_online.yaml [hunk 2]
```diff
@@ -1,3 +1,3 @@
 mode: "online"
-entity: "viper_svg"
-project: "reppo"
+entity: "idutta17-ut-austin"
```
Comment (cvoelcker) [inline config/logging/wandb_online.yaml]: Same here, I think the best default would actually be to leave this blank. Otherwise everybody else who uses this code will get errors.

File: src/algorithms/reppo/reppo_policy.py [hunk 3]
```diff
@@ -27,7 +27,7 @@ def __init__(
     ):
         self.base = base
         self.normalizer = normalizer
-        self.normalization_state = normalization_state
+        self.normalization_state = nnx.data(normalization_state) if normalization_state is not None else None
```
Comment (cvoelcker) [inline src/algorithms/reppo/reppo_policy.py:30]: Why were these changes necessary?

File: src/algorithms/envs.py [hunk 4]
```diff
@@ -187,11 +205,35 @@ def make_env(eval: bool = False):
 
     env = make_env(eval=False)
     eval_env = make_env(eval=True)
+
```
Comment (cvoelcker) [inline src/algorithms/envs.py:209]: Please remove debug logging, or change the log level from info to debug

File: src/algorithms/envs.py [hunk 5]
```diff
@@ -187,11 +205,35 @@ def make_env(eval: bool = False):
...
     eval_env = make_env(eval=True)
+    
+    if cfg.algorithm.bc_indicator:
+        # Debug: check observation space from wrapper
+        logging.info(f"[make_maniskill_env] env.single_observation_space: {env.single_observation_space}")
+        logging.info(f"[make_maniskill_env] env.single_observation_space.shape: {env.single_observation_space.shape}")
+        logging.info(f"[make_maniskill_env] env.observation_space: {env.observation_space}")
+        logging.info(f"[make_maniskill_env] env.observation_space.shape: {env.observation_space.shape}")
+        
+        # Override observation space to match dataset dimension
```
Comment (cvoelcker) [inline src/algorithms/envs.py:211]: I'm confused: is this safe to do? Shouldn't we make sure that the env observation space, and the dataset agree?

File: src/algorithms/training.py [hunk 6]
```diff
@@ -171,11 +182,18 @@ def loop_train_fn(key: Key) -> tuple[TrainState, dict]:
         key, init_key = jax.random.split(key)
         state = init_fn(init_key)
         obs, _ = env.reset()
-        state = state.replace(last_obs=to_jax(obs), last_env_state=None)
+        state = state.replace(last_obs=obs, last_env_state=None)
         logging.info(f"Starting training for {num_iterations} iterations.")
         logging.info(f"Train steps per iteration: {train_steps_per_iteration}.")
         logging.info(f"Total time steps: {total_time_steps}.")
 
+        # Create checkpoint directories under src/ for saving model parameters
```
Comment (cvoelcker) [inline src/algorithms/training.py]: This should best case be configurable by config and not be tied to src/

Comment (cvoelcker) [inline src/algorithms/training.py]: In general, no run data should be placed under src, src is for code source ;)

File: src/algorithms/utils.py [hunk 7]
```diff
@@ -197,19 +199,45 @@ def print_logs(metrics):
...
-        _env_state = env_state.unwrapped()
-        key, randomize_steps_key = jax.random.split(key)
-        _env_state.info["steps"] = jax.random.randint(
-            randomize_steps_key,
-            _env_state.info["steps"].shape,
-            0,
-            env.episode_length,
-        ).astype(jnp.float32)
-        env_state.set_env_state(_env_state)
+    # Unsure about this part - this particular envionment configuration doesn't really behave the same way as the other one and it does not have the 'step' attribute
```
Comment (cvoelcker) [inline src/algorithms/utils.py]: I don't understand the question, but this might be something we should talk about?

File: src/pretrain-torch.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+        # NLL: Maximum likelihood estimation
+        nll_loss = -log_prob.sum(dim=-1).mean()
+        nll_loss_sum += nll_loss.item()
+        
+        # MSE loss
+        mse_loss = (tanh_mean - expert_action).pow(2).mean()
+        mse_loss_sum += mse_loss.item()
+
+        # Combined actor loss
+        actor_loss = nll_loss + cfg.algorithm.bc_mse_loss_weight * mse_loss  # - std_reg_coef * log_std.mean()
```
Comment (cvoelcker) [inline src/pretrain-torch.py:56]: Have you checked that you can still train this and switch the MSE loss off?

File: src/pretrain-torch.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+    # Initialize wandb
+    wandb.init(
+        config=dict(cfg),
+        entity=cfg.logging.entity,
+        project=cfg.logging.project,
+        name=f"bc_pretrain_no_clamp_linear_scale_no_envstates_{cfg.env.name}",
+        mode=cfg.logging.mode
+    )
+    
+    global device, train_loader, val_loader, optimizer
```
Comment (cvoelcker) [inline src/pretrain-torch.py:97]: I really, really, really dislike the use of global. Please avoid unless absolutely necessary.

File: src/pretrain-torch.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    
+    # Initializing in a separate cell so we can easily add more epochs to the same run
+    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
+    writer = SummaryWriter('runs/fashion_trainer_{}'.format(timestamp))
+    epoch_number = 0
+
+    # Initilaize BC specific variables
+    EPOCHS = 128
```
Comment (cvoelcker) [inline src/pretrain-torch.py:105]: That's a lot of epochs, also, this should probably be in a config somewhere

File: src/pretrain-torch.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+        log_std_sum += log_std.mean().item()
+
+        log_prob = dist.log_prob(expert_action)  # [B, action_dim]
+        
+        # NLL: Maximum likelihood estimation
+        nll_loss = -log_prob.sum(dim=-1).mean()
+        nll_loss_sum += nll_loss.item()
+        
+        # MSE loss
+        mse_loss = (tanh_mean - expert_action).pow(2).mean()
```
Comment (cvoelcker) [inline src/pretrain-torch.py:52]: There is some potential weirdness here about the interaction of tanh and mean. This can have effects. The tanh mean pre squashed gaussian is not the mean of the distribution. We normally ignore this, but if you explicitly train with it, that might not work well

File: src/pretrain-torch.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+    print(f"Expert action min (raw): {expert_action_min:.6f}")
+    print(f"Expert action max (raw): {expert_action_max:.6f}")
+    print(f"Normalized action min: {normalized_action_min:.6f}")
+    print(f"Normalized action max: {normalized_action_max:.6f}")
+    print(f"Action normalization bounds (low): {low.cpu().numpy()}")
+    print(f"Action normalization bounds (high): {high.cpu().numpy()}")
+    print("================================\n")
+    time = datetime.now().strftime("%Y%m%d_%H%M%S")
+
+    for epoch in range(EPOCHS):
```
Comment (cvoelcker) [inline src/pretrain-torch.py:191]: I think there is code repetition here with the train_one_epoch above?

**Review by cvoelcker — COMMENTED** [review]
I don't see the logic for applying the data normalization in the RL finetuning part. Can you point that out to me?

File: src/runners/maniskill_runner.py [hunk 13]
```diff
@@ -15,79 +20,284 @@
...
+            obs = flatten_obs(obs_dict, env=env, demo_obs_keys=demo_obs_keys)
+            
+            prev_step = train_state.time_steps
+            prev_time = time.perf_counter()
+            for i in range(num_steps):
+                key, act_key = jax.random.split(key)
+                action, _ = policy(act_key, obs)
+                
+                # Convert action from JAX to numpy for environment
+                action = np.asarray(action)
```
Comment (cvoelcker) [inline src/runners/maniskill_runner.py]: Casting stuff via numpy will make things very slow potentially

File: src/runners/maniskill_runner.py [hunk 14]
```diff
@@ -15,79 +20,284 @@
...
+                    _next_obs = to_jax(flatten_obs(info["final_observation"], env=env, demo_obs_keys=demo_obs_keys))
+                else:
+                    _next_obs = flatten_obs(next_obs_dict, env=env, demo_obs_keys=demo_obs_keys)
+            
+                # Convert torch tensors to numpy, handling CUDA
+                reward = torch_to_numpy(reward)
+                done = torch_to_numpy(done)
+                truncated = torch_to_numpy(truncated)
+                # Normalize action back to [-1, 1] for storage in transitions
+                action = normalize_action(action, dataset_low, dataset_high)
```
Comment (cvoelcker) [inline src/runners/maniskill_runner.py]: One sanity check test I would run is whether denormalize(normalize(action)) gives you back the correct action. Double check that

File: src/runners/maniskill_runner.py [hunk 15]
```diff
@@ -15,79 +20,284 @@
...
+    
+    if not obs_list:
+        raise ValueError(f"No observation data found in obs_dict with keys: {obs_dict.keys()}")
+    
+    result = np.concatenate(obs_list, axis=1)
+    return result
+
+def make_rollout_fn(env: gymnasium.Env, num_steps: int, num_envs: int, demo_path: str = None, bc_indicator: bool = False) -> RolloutFn:
+    # BC-specific rollout function with demo observation flattening
+    if bc_indicator:
```
Comment (cvoelcker) [inline src/runners/maniskill_runner.py]: Is this only true for BC training, or for all training with BC pretraining?

File: src/algorithms/reppo/ff_reppo.py [hunk 16]
```diff
@@ -21,10 +23,247 @@
...
+        
+        logging.info(f"Found {len(torch_linear_layers)} linear layers in torch model")
+        
+        # Get JAX actor's feature_encoder structure
+        fe = jax_actor.feature_encoder
+        
+        # Collect JAX linear layers (from input_layer, main_layers, output_layer)
+        jax_linear_refs = []
+        
+        # input_layer: Sequential with [Linear, LayerNorm(?), Activation]
```
Comment (cvoelcker) [inline src/algorithms/reppo/ff_reppo.py]: Could it be that you didn't translate the layer norms??? I think that might be the issue?

File: src/algorithms/reppo/reppo_policy.py (hunk 3 above)
Comment (indira-17) [reply src/algorithms/reppo/reppo_policy.py:30]: Otherwise, it produces the following error:

```
This is an error starting from Flax version 0.12.0.
Consider one of the following options:
If the attribute is meant to be static, either remove the Array value or wrap it in a static container.
Wrap the value with nnx.data on assignment:
  _.normalization_state = nnx.data(...)
Annotate the class attribute with nnx.data:
class REPPOPolicy(Module):
normalization_state: NormalizationState = nnx.data()
If the container is a list or dict, try using nnx.List(...) or nnx.Dict(...) instead.
Disable pytree for this class:
class REPPOPolicy(Module, pytree=False):
```

File: src/algorithms/envs.py (hunk 4 above)
Comment (indira-17) [reply src/algorithms/envs.py:209]: Changed to logging.debug

File: src/algorithms/envs.py (hunk 5 above)
Comment (indira-17) [reply src/algorithms/envs.py:211]: I don't think we would need it anymore after the latest fixes

File: src/runners/maniskill_runner.py (hunk 13 above)
Comment (indira-17) [reply src/runners/maniskill_runner.py]: The if bc_indicator=True might not be needed anymore but will have to test it before removing.

File: src/runners/maniskill_runner.py (hunk 14 above)
Comment (indira-17) [reply src/runners/maniskill_runner.py]: I checked a couple of tasks, and the total difference across dimensions and trajectories is on the order of 6x10^(-8) to 8x10^(-8).

**Review by indira-17 — COMMENTED** [review]
Final fixes after incorporating some of the feedbacks
